File: nodejs/src/exceptions.ts
/**
 * Custom exception hierarchy for DatabricksWorkspaceClient.
 *
 * Each exception supports ES2022 Error.cause for chaining the original error,
 * mirroring Python's `raise ... from` pattern.
 */

export class DatabricksClientError extends Error {
  readonly details: Record<string, unknown>;

  constructor(
    message: string,
    options?: { details?: Record<string, unknown>; cause?: unknown },
  ) {
    super(message, { cause: options?.cause });
    this.name = "DatabricksClientError";
    this.details = options?.details ?? {};
  }
}

export class AuthenticationError extends DatabricksClientError {
  constructor(
    message: string,
    options?: { details?: Record<string, unknown>; cause?: unknown },
  ) {
    super(message, options);
    this.name = "AuthenticationError";
  }
}

export class ResourceNotFoundError extends DatabricksClientError {
  constructor(
    message: string,
    options?: { details?: Record<string, unknown>; cause?: unknown },
  ) {
    super(message, options);
    this.name = "ResourceNotFoundError";
  }
}

export class PermissionDeniedError extends DatabricksClientError {
  constructor(
    message: string,
    options?: { details?: Record<string, unknown>; cause?: unknown },
  ) {
    super(message, options);
    this.name = "PermissionDeniedError";
  }
}

export class QueryExecutionError extends DatabricksClientError {
  constructor(
    message: string,
    options?: { details?: Record<string, unknown>; cause?: unknown },
  ) {
    super(message, options);
    this.name = "QueryExecutionError";
  }
}

export class JobRunError extends DatabricksClientError {
  constructor(
    message: string,
    options?: { details?: Record<string, unknown>; cause?: unknown },
  ) {
    super(message, options);
    this.name = "JobRunError";
  }
}

export class OperationTimeoutError extends DatabricksClientError {
  constructor(
    message: string,
    options?: { details?: Record<string, unknown>; cause?: unknown },
  ) {
    super(message, options);
    this.name = "OperationTimeoutError";
  }
}

export class AmbiguousJobError extends DatabricksClientError {
  constructor(
    message: string,
    options?: { details?: Record<string, unknown>; cause?: unknown },
  ) {
    super(message, options);
    this.name = "AmbiguousJobError";
  }
}

File: python/src/databricks_workspace_client/exceptions.py
"""Custom exception hierarchy for DatabricksWorkspaceClient.

Each exception wraps the original SDK exception as __cause__ via `raise ... from`,
preserving full debuggability while providing a cross-language-consistent error interface.
"""

from __future__ import annotations


class DatabricksClientError(Exception):
    """Base exception for all DatabricksWorkspaceClient errors."""

    def __init__(self, message: str, *, details: dict[str, object] | None = None) -> None:
        super().__init__(message)
        self.details = details or {}


class AuthenticationError(DatabricksClientError):
    """Failed to authenticate with the Databricks workspace."""


class ResourceNotFoundError(DatabricksClientError):
    """Requested resource (job, run, warehouse, etc.) was not found."""


class PermissionDeniedError(DatabricksClientError):
    """Caller lacks permission for the requested operation."""


class QueryExecutionError(DatabricksClientError):
    """SQL statement execution failed."""


class JobRunError(DatabricksClientError):
    """Job run failed or reached an unexpected terminal state."""


class OperationTimeoutError(DatabricksClientError):
    """Operation exceeded the configured timeout."""


class AmbiguousJobError(DatabricksClientError):
    """Job name matched multiple jobs. Provide a more specific name or use job_id."""
